### DISCUSSION TITLE
Using @collection within create API Rule to limit amount of records a user can create

### DISCUSSION THREAD

**Discussion by abrman** [pocketbase/pocketbase#3108] [Q&A] [answered]
I'm trying to limit how many projects a user can create in `projects` collection based on `user.project_limit`.
To achieve this I have a View Collection `v_user_projects_counts` where it counts the amount of projects a user has created.
```sql
SELECT u.id, COUNT(p.id) AS project_count, u.project_limit
FROM users u
LEFT JOIN projects p ON u.id = p.created_by
GROUP BY u.id;
```

I think I'm struggling with understanding how to build the create rule. I'd assume this would work.
```
@request.auth.id = created_by.id  &&
@request.auth.id = @collection.v_user_projects_counts.id && 
@collection.v_user_projects_counts.project_count <= @collection.v_user_projects_counts.project_limit
```
The idea is to verify that the project a user is creating is referencing himself in `created_by.id`, LEFT JOIN on `request.auth.id` = `v_user_projects_counts.id`, then make sure that the `v_user_projects_counts.project_count` is less than or equal to `v_user_projects_counts.project_limit` allowing him to create a new project.

It seemed to work with one user, however when I introduced a second user it started to behave in an unintended way. 
```js
await pb.collection("projects").create({
  name: `New Project`,
  ...
  created_by: user.id,
});
```
results in error found in logs:
```json
{
  "errorDetails": "DrySubmit create rule failure: sql: no rows in result set",
  "errorMessage": "Failed to create record."
}
```

Also, is my understanding correct that `@request.data.created_by.id` and `created_by.id` are the same for Create API rule?

Screenshots for context:
![image](https://github.com/pocketbase/pocketbase/assets/6663260/35361928-4a34-4718-8785-0760f9c3091e)
![image](https://github.com/pocketbase/pocketbase/assets/6663260/8b8a3770-9692-4718-8841-cdacacbe069e)

**Comment by ganigeorgiev** [comment, marked as answer]
The approach seems ok with one minor details - it seems that you want "at least one/any" join (aka. use the `?` prefix operator with the `@collection.*` filter, otherwise it will require every joined collection record to match the condition).

If I'm understanding correctly your structure, the rule should be something like:

```js
@request.auth.id = created_by  &&
@collection.v_user_projects_counts.id ?= @request.auth.id && 
@collection.v_user_projects_counts.project_count ?<= @collection.v_user_projects_counts.project_limit
```
(the last part could be also written as `@collection.v_user_projects_counts.project_count ?<= @request.auth.project_limit`)


> Also, is my understanding correct that @request.data.created_by.id and created_by.id are the same for Create API rule?

Yes. This is because the Create API rule is applied **after** a dry transactional submit (_if it fails we rollback the transaction_).

**Reply by abrman** [reply]
Works perfectly! <3
I did try "at least one/any" join on the id, but then both users were affected by limit of 2 which was imposed on just one user.
```
@request.auth.id = created_by.id  &&
@request.auth.id ?= @collection.v_user_projects_counts.id && 
@collection.v_user_projects_counts.project_count <= @collection.v_user_projects_counts.project_limit
```
Think my SQL brain isn't present. 
How does `?<=` work with in connection with the `@collection.v_user_projects_counts.id ?= @request.auth.id`

I understand that `@collection.v_user_projects_counts.id ?= @request.auth.id` is satisfied since at least one row was found. How does `@collection.v_user_projects_counts.project_count ?<= @collection.v_user_projects_counts.project_limit` know to check the satisfied row from the previous check? And if it checks against the ones that were filtered by the previous check, why didn't `<=` work, and has to be `?<=` to satisfy the logic I'm going for?

Thank you for taking the time to help me figure this out!

**Reply by ganigeorgiev** [reply]
When `?` is used in both `@collection.*` condition, the above rule will be roughly translated to a SQL query like:

```sql
SELECT projects.*
FROM projects
LEFT JOIN v_user_projects_counts as v
WHERE projects.created_by = {:authId} AND v.id = {:authId} AND v.project_count <= v.project_limit
```
(_essentially the rule represent the `WHERE` condition_)

When a match all expression is used (aka. when there is no `?` prefix with the joined records or multiple relations) the generated SQL will be similar to the above but with extra **subquery(ies)** that will check whether all records matches the desired condition(s).

I agree that the `?` syntax is a little confusing (_I have it in my todo to think on a different and less ambiguous notation_). It was introduced in [v0.11.0](https://github.com/pocketbase/pocketbase/releases/tag/v0.11.0) and I deliberately decided to make the match-all behavior the default one because it is the safer option most of the time (_this is especially true when submitting for example a multiple relation field and want to ensure that **all** of the newly submitted relations are owned by the current user id, aka. `someMultipleRel.user = @request.auth.id`_).

**Comment by abrman** [comment]
Hey @ganigeorgiev.

I'm following up (after like 6 months 👀) due to this comment,
> Yes. This is because the Create API rule is applied after a dry transactional submit (if it fails we rollback the transaction).

I have implemented versioning on beforementioned `projects` by having a `project` with relation to `project_data` where bulk of the project data lives. Is there any way in the rules to prevent the last two project_datas from being deleted?

I initially tried a view collection with the following SQL that keeps the last two IDs I'd like to prevent from being deleted
```sql
SELECT 
  (ROW_NUMBER() OVER() || '') as id,
  p.id as project_id,
  JSON_ARRAY(
    JSON_EXTRACT(p.project_datas, '$[' || (JSON_ARRAY_LENGTH(p.project_datas) - 2) || ']'),
    JSON_EXTRACT(p.project_datas, '$[' || (JSON_ARRAY_LENGTH(p.project_datas) - 1) || ']')
  ) AS no_delete_ids
FROM
  Projects p
``` 
however the whole dry submit got back at me. Any ideas how to achieve such functionality?

My `project_data` table delete rule is following:
```
@request.auth.id != "" && 
@request.auth.projects.project_datas.id ?!= @collection.v_projects_project_datas_no_delete.no_delete_ids
```

**Reply by ganigeorgiev** [reply]
I don't understand what is the desired behavior.
There is no dry submit for the delete action. On delete we check the delete API rule and if it resolves to true we proceed with the db delete of the record.

**Reply by ganigeorgiev** [reply]
Also note that the View collection is "dynamic", aka. there is no persistence of the View collection data and the view query is executed every time when accessed which is maybe what causes your rule to fail but again I don't understand what is the desired behavior and what the above rule is supposed to check.

**Reply by ganigeorgiev** [reply]
Also #2, ignoring the way how you construct the `JSON_ARRAY` expression, if `no_delete_ids` is a json array we'll not iterate over its values if that is what you are trying to do (_the [`:each` modifier](https://pocketbase.io/docs/api-rules-and-filters/#each-modifier) currently works only with `select` field type_).

**Reply by abrman** [reply]
I just wanted to prevent deletion of all project data a user may have worked on, either by a bug, or intention.
`users` have a relation to `projects` (projects they've created)
`projects` have a relation to `project_datas`
`project_datas` is like a version history and you always display from the top of stack.

I'm currently making sure old ones get deleted as they're 99% of the time not necessary and as projects grew over time, they end up taking quite a bit of space.

Ideally you should be unable to delete from the top of the stack (latest version) that was my intention initially by listing the IDs and trying to make sure they don't get included in the `no_delete_ids` list, however I resolved it with a view collection that just counts how many `project_datas` the user has and prevents him from deleting last three (due to possible corruption, coffee spills, etc.)

I did it much the same way as I did earlier in this thread

Delete rule:
```
@request.auth.id != "" && 
@request.auth.projects.project_datas.id ?= id && 
@collection.v_projects_project_datas_count.project_id ?= @request.auth.projects.id && 
@collection.v_projects_project_datas_count.project_datas_count ?> 3
```

`v_project_datas_count`
```sql
SELECT 
  (ROW_NUMBER() OVER() || '') as id,
  p.id as project_id,
  JSON_ARRAY_LENGTH(p.project_datas) AS project_datas_count
FROM
  Projects p
```

Thank you @ganigeorgiev for continued development and support of pocketbase. it wouldn't be the same without your commitment.
I'm dumping my resolution in case someone stumbles across this thread in the future.
